File: Raghav Bhardwaj/frontend/src/utils/roles.js
/**
 * Role constants — must match backend app/rbac/roles.py
 *
 * CHANGES: Removed AUDITOR role. Merged REVIEWER and APPROVER into single APPROVER role.
 * The APPROVER role can now:
 *   • Review submissions and check evidence (validates completeness)
 *   • Approve / Return / Escalate after reviewing
 * Only 4 roles remain: ADMIN, PREPARER, APPROVER, CERTIFIER
 */

export const ROLES = {
  ADMIN:     'admin',
  PREPARER:  'preparer',
  APPROVER:  'approver',
  CERTIFIER: 'certifier',
}

/** Normalise raw role string to lowercase — NO cross-role aliasing. */
export const normalizeRole = (role) => (role || '').toLowerCase().trim()

/** True if the role can perform write / state-change operations. */
export const isWriteRole = (role) =>
  [ROLES.ADMIN, ROLES.PREPARER, ROLES.APPROVER, ROLES.CERTIFIER].includes(
    normalizeRole(role)
  )

/** True if the role is strictly read-only (none — auditor role removed). */
export const isReadOnlyRole = (role) => false

/** True if the role has approver-or-above authority (approver, certifier, admin). */
export const isApproverOrAbove = (role) =>
  [ROLES.APPROVER, ROLES.CERTIFIER, ROLES.ADMIN].includes(normalizeRole(role))

File: Raghav Bhardwaj/backend/app/rbac/roles.py
ADMIN     = "admin"
PREPARER  = "preparer"
APPROVER  = "approver"
CERTIFIER = "certifier"

ALL_ROLES = {ADMIN, PREPARER, APPROVER, CERTIFIER}

# ---------------------------------------------------------------------------
# Role hierarchy (higher index = more authority).
# Used by SoD checks to ensure approvals always move UP the chain.
# APPROVER combines review + approval in a single role that can:
#   • Review submissions and check evidence
#   • Approve / return / escalate after reviewing the preparer's work
# ---------------------------------------------------------------------------
ROLE_RANK: dict[str, int] = {
    PREPARER:  1,
    APPROVER:  2,   # merged reviewer & approver — handles both review and approval
    CERTIFIER: 3,
    ADMIN:     99,
}

# Roles that are permitted to make write / state-change calls
WRITE_ROLES = {ADMIN, PREPARER, APPROVER, CERTIFIER}

# Roles that are strictly read-only on data endpoints (none — auditor role removed)
READ_ONLY_ROLES = set()
